docker-app/qfieldcloud/filestorage/utils.py

```
@@ -155,3 +156,27 @@ def to_uuid(value: Any) -> uuid.UUID | None:
         return uuid.UUID(value)
     except ValueError:
         return None
+
+
+def parse_range(input_range: str) -> tuple[int, int | None] | None:
```

Would you mind adding a few unit tests for this? I think it is quite easy to test. 

I am also wondering what the RFC says when you send invalid request?

If one sends:
```
Range: third-10e6
```
You will ignore it because it will not match the regex, but if someone sends:
```
Range: 10-5
```

You will raise?

Shall you ignore the header also?

If yes, then you better move the `if` from:
`if (
            range_start >= file_size
            or range_end >= file_size
            or range_end < range_start
        ):
```
here, it feels more natural to me.

---

> I am also wondering what the RFC says when you send invalid request?

A question I'm also asking myself, so far:

- multiple ranges are not supported at the moment
-  [RFC](https://www.rfc-editor.org/rfc/rfc9110.html#name-range) says: "A server that supports range requests MAY ignore or reject a [Range](https://www.rfc-editor.org/rfc/rfc9110.html#field.range) header field that contains an invalid [ranges-specifier](https://www.rfc-editor.org/rfc/rfc9110.html#rule.ranges-specifier) ([Section 14.1.1](https://www.rfc-editor.org/rfc/rfc9110.html#range.specifiers)), a [ranges-specifier](https://www.rfc-editor.org/rfc/rfc9110.html#rule.ranges-specifier) with more than two overlapping ranges, or a set of many small ranges that are not listed in ascending order, since these are indications of either a broken client or a deliberate denial-of-service attack ([Section 17.15](https://www.rfc-editor.org/rfc/rfc9110.html#overlapping.ranges))" -> not sure I'm reading this correctly, but it seems to me that each implementation may ignore or reject invalid range header

- [RFC](https://www.rfc-editor.org/rfc/rfc9110.html#section-14.1.2-9.1.1) says: "The final 500 bytes (byte offsets 9500-9999, inclusive): bytes=-500" -> should we make the code compliant in this PR ?

---

> if someone sends:

> `Range: 10-5`

> You will raise?

> Shall you ignore the header also?

I think we should answer to that question : ignore or reject. Current state is that invalid requests are rejected, my personal opinion favors this approach, since the server will not get an overload in this case.

---

1) multipart - totally fine if not supported, just add a test to confirm they are not supported. Also probably add it to the `parse_range` function doc string. You may also prepare the function to return a list of ranges already, you can assert in the code it returns only 1 range.
2) yes, I read the same, it's up to the implementation to decide whether to ignore or reject the header. I like being vocal rather than silent, so we better always return an error when we are not happy with the header.
3) I don't think it is needed, unless you say it is very easy to achieve so and it will work fine with Django content file. If someone wants this supported, we are open for PRs :)
4) Reject whenever possible.